lib/puppet/network/http/connection.rb

```diff
@@ -189,6 +189,31 @@ def request_with_redirects(request, options)
             request.each do |header, value|
               current_request[header] = value
             end
+          when 503, 429
+            retry_after = current_response['Retry-After']
+
+            if retry_after.nil?
+              # Set return value and advance to next loop.
+              response = current_response
+              next
+            end
+
+            begin
+              retry_sleep = Integer(retry_after)
+            rescue TypeError, ArgumentError
+              Puppet.err(_("Recieved a %{status_code} response from the server, but the Retry-After header value of '%{retry_after}' could not be converted to an integer.") %
+                         {status_code: current_response.code,
+                          retry_after: retry_after.inspect})
+
+              response = current_response
+              next
+            end
+
+            Puppet.warning(_('Recieved a %{status_code} response from the server. Sleeping for %{retry_after} seconds before retrying the request.') %
+                           {status_code: current_response.code,
+                            retry_after: retry_after})
+
+            sleep(retry_sleep)
```

Comment:
Should we set a max number of times we will try to sleep/delay so we don't get into an infinite loop?